File: 2020/cpp/day_14.cpp
#include "day_14.h"

day_14::computer_t::computer_t(const std::vector<instruction_t>& input) : input{input}, instruction_pointer{input.begin()} {
}

bool day_14::computer_t::nextInstruction() {
    switch (instruction_pointer->type) {
        case instruction_t::Mask:
            mask = instruction_pointer->mask;
            break;
        case instruction_t::Mem:
            const auto zeroed = instruction_pointer->value & (~mask.zeros);
            const auto masked = zeroed | mask.ones;
            memory[instruction_pointer->address] = masked;
            break;
    }
    ++instruction_pointer;
    if (instruction_pointer == input.end()) {
        return false;
    }
    return true;
}
void day_14::computer_t::reset() {
    instruction_pointer = input.begin();
    memory.clear();
}

day_14::day_14() :
    input{[] {
        const auto input_strings = readFile("Inputs/day_14.txt");
        const auto start = std::chrono::high_resolution_clock::now();
        std::vector<computer_t::instruction_t> input;
        for (const auto& line : input_strings) {
            auto& instruction = input.emplace_back();
            switch (line[1]) {
                case 'a': {  // mask instruction
                    instruction.type = computer_t::instruction_t::Mask;
                    auto* const mask_str = line.data() + 7;
                    for (int i = 35; i >= 0; --i) {
                        switch (mask_str[i]) {
                            case '1':
                                instruction.mask.ones |= 1ull << (35 - i);
                                break;
                            case '0':
                                instruction.mask.zeros |= 1ull << (35 - i);
                                break;
                            case 'X':
                                instruction.mask.xs |= 1ull << (35 - i);
                                break;
                        }
                    }
                    // fmt::print("str:\t{}\nones:\t{:0>36b}\nzeroes:\t{:0>36b}\nXs:\t\t{:0>36b}\n", mask_str, instruction.mask.ones, instruction.mask.zeros,
                    //            instruction.mask.xs);
                    break;
                }
                case 'e':  // mem instruction
                    instruction.type = computer_t::instruction_t::Mem;
                    auto splitted = split(line, '=');
                    instruction.address = fast_atol(line.data() + 4);
                    instruction.value = fast_atol(splitted.back().data() + 1);
                    break;
            }
        }
        const auto end = std::chrono::high_resolution_clock::now();
      fmt::print("Parsing input for day fourteen took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
        return input;
    }()},
    computer{input} {
}

void day_14::part_one() {
    const auto start = std::chrono::high_resolution_clock::now();
    computer.reset();
    while (computer.nextInstruction()) {
    }
    uint64_t count{0};
    for (auto [addr, val] : computer.memory) {
        count += val;
    }
    auto end = std::chrono::high_resolution_clock::now();
    fmt::print("The answer for day fourteen part one is {}\n", count);
    fmt::print("Took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

void day_14::part_two() const {
}

File: 2020/Rust/src/day14.rs
use std::{
    collections::HashMap,
    time::{Duration, Instant},
};

#[derive(Debug)]
enum Instruction {
    Mem { location: usize, value: usize },
    Mask(Mask_),
}

#[derive(Debug, Default, Clone)]
struct Mask_ {
    ones: usize,
    zeros: usize,
    xs: Vec<usize>,
}

fn parse_input(input: &str) -> (Vec<Instruction>, Duration) {
    let start = Instant::now();
    let out = input
        .lines()
        .map(|line| {
            if line.starts_with("mask =") {
                let mut ones = 0;
                let mut zeros = 0;
                let mut xs = Vec::new();
                let s = line[7..].as_bytes();
                for (i, c) in s.iter().rev().enumerate() {
                    match c {
                        b'0' => zeros |= 1 << i,
                        b'1' => ones |= 1 << i,
                        b'X' => xs.push(i),
                        _ => unreachable!("invalid input"),
                    }
                }
                Instruction::Mask(Mask_ { ones, zeros, xs })
            } else if line.starts_with("mem[") {
                let location = line[4..line.find(']').unwrap()].parse().unwrap();
                let value = line[line.find('=').unwrap() + 2..].parse().unwrap();
                Instruction::Mem { location, value }
            } else {
                unreachable!("invalid input");
            }
        })
        .collect();
    /*
     *let mut out = Vec::with_capacity(input.lines().count());
     *for line in input.lines() {
     *    if line.starts_with("mask =") {
     *        let mut ones = 0;
     *        let mut zeros = 0;
     *        let mut xs = Vec::new();
     *        let s = line[7..].as_bytes();
     *        for (i, c) in s.iter().rev().enumerate() {
     *            match c {
     *                b'0' => zeros |= 1 << i,
     *                b'1' => ones |= 1 << i,
     *                b'X' => xs.push(i),
     *                _ => unreachable!("invalid input"),
     *            }
     *        }
     *        out.push(Instruction::Mask(Mask_ { ones, zeros, xs }));
     *    } else if line.starts_with("mem[") {
     *        let location = line[4..line.find(']').unwrap()].parse().unwrap();
     *        let value = line[line.find('=').unwrap() + 2..].parse().unwrap();
     *        out.push(Instruction::Mem { location, value });
     *    } else {
     *        unreachable!("invalid input");
     *    }
     *}
     */
    let end = Instant::now();
    (out, end - start)
}

fn solve_part1(input: &[Instruction]) -> (usize, Duration) {
    let start = Instant::now();
    let mut memory = HashMap::new();
    let mut mask_zeros = 0;
    let mut mask_ones = 0;

    for instruction in input {
        match instruction {
            &Instruction::Mem { location, value } => {
                let val = value & (!mask_zeros);
                let val = val | mask_ones;
                memory.insert(location, val);
            }
            Instruction::Mask(mask) => {
                mask_ones = mask.ones;
                mask_zeros = mask.zeros;
            }
        };
    }
    let count = memory.values().sum();
    let end = Instant::now();
    (count, end - start)
}

fn solve_part2(input: &[Instruction]) -> (usize, Duration) {
    let start = Instant::now();
    let mut memory = HashMap::new();
    let mut floating = Vec::new();
    let mut whitelist = 0;

    for instruction in input {
        match instruction {
            &Instruction::Mem { location, value } => {
                let addr = location & whitelist;
                for float_addr in &floating {
                    memory.insert(addr | float_addr, value);
                }
            }
            Instruction::Mask(mask) => {
                whitelist = mask.zeros;
                floating.clear();
                for template in 0..2usize.pow(mask.xs.len() as u32) {
                    let addr = mask.xs.iter().enumerate().fold(mask.ones, |addr, (i, fb)| {
                        addr | (template & 1 << i) << (fb - i)
                    });
                    floating.push(addr);
                }
            }
        };
    }

    let count = memory.values().sum();
    let end = Instant::now();
    (count, end - start)
}

pub fn run(input: String) {
    let (parsed_input, time) = parse_input(&input);
    println!("Day 14, parsing input took {}ns", time.as_nanos());
    let (p1, time) = solve_part1(&parsed_input);
    println!("Day 14, part 1: {}. Took {}ns", p1, time.as_nanos());
    let (p2, time) = solve_part2(&parsed_input);
    println!("Day 14, part 2: {}. Took {}ns", p2, time.as_nanos());
}
